apstra/api_custom_collector.go

```diff
@@ -0,0 +1,286 @@
+	// Check if this is the first collector for that service name
+	//cs, err := o.GetCollectorsByServiceName(ctx, in.ServiceName)
+	//if err != nil {
+	//	return err
+	//}
+	var Request struct {
+		ServiceName string      `json:"service_name"`
+		Items       []Collector `json:"collectors"`
+	}
+	Request.ServiceName = in.ServiceName
+	Request.Items = append(Request.Items, *in)
+
+	lockId := fmt.Sprintf(CollectorLock, in.ServiceName)
+	o.lock(lockId)
+	defer o.unlock(lockId)
+
+	// This is the first collector for this service name
+	// So we POST
+	err := o.talkToApstra(ctx, &talkToApstraIn{
+		method:   http.MethodPost,
+		urlStr:   fmt.Sprintf(apiUrlCollectors),
+		apiInput: &Request,
+	})
+	err = convertTtaeToAceWherePossible(err)
+	var ace ClientErr
+	if !(errors.As(err, &ace) && ace.Type() == ErrConflict) {
+		return err // fatal error
+	}
+
+	// There are other collectors, so this is a patch
+	return convertTtaeToAceWherePossible(o.talkToApstra(ctx, &talkToApstraIn{
+		method:   http.MethodPatch,
+		urlStr:   fmt.Sprintf(apiUrlCollectorsByServiceName, in.ServiceName),
+		apiInput: &Request,
+	}))
+}
+
+// UpdateCollector Updates a collector
+func (o *Client) UpdateCollector(ctx context.Context, in *Collector) error {
+	var Request struct {
+		Collectors []Collector `json:"collectors"`
+	}
+	Request.Collectors = append(Request.Collectors, *in)
+	return convertTtaeToAceWherePossible(o.talkToApstra(ctx, &talkToApstraIn{
+		method:   http.MethodPatch,
+		urlStr:   fmt.Sprintf(apiUrlCollectorsByServiceName, in.ServiceName),
+		apiInput: &Request,
+	}))
+}
+
+// DeleteAllCollectorsInService deletes all the collectors under a service
+func (o *Client) DeleteAllCollectorsInService(ctx context.Context, name string) error {
+	return convertTtaeToAceWherePossible(o.talkToApstra(ctx, &talkToApstraIn{
+		method: http.MethodDelete,
+		urlStr: fmt.Sprintf(apiUrlCollectorsByServiceName, name),
+	}))
+}
+
+func (p1 *CollectorPlatform) Equals(p2 *CollectorPlatform) bool {
+	if p1.OsType != p2.OsType {
+		return false
+	}
+	if p1.OsVersion != p2.OsVersion {
+		return false
+	}
+	if p1.Model != p2.Model {
+		return false
+	}
+	if len(p1.OsFamily) != len(p2.OsFamily) {
+		return false
+	}
+
+	m := make(map[CollectorOSFamily]bool, len(p1.OsFamily))
+	for _, v := range p1.OsFamily {
+		m[v] = true
+	}
+	for _, v := range p2.OsFamily {
+		_, ok := m[v]
+		if !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// DeleteCollector deletes the collector described in the object
+func (o *Client) DeleteCollector(ctx context.Context, in *Collector) error {
+	var Request struct {
+		ServiceName string      `json:"service_name"`
+		Items       []Collector `json:"collectors"`
+	}
+
+	lockId := fmt.Sprintf(CollectorLock, in.ServiceName)
+	o.lock(lockId)
+	defer o.unlock(lockId)
+
+	cs, err := o.GetCollectorsByServiceName(ctx, in.ServiceName)
+	if err != nil {
+		return convertTtaeToAceWherePossible(err)
```

Comment:
There shouldn't be any harm in invoking `convertTtaeToAceWherePossible` here, but it shouldn't be necessary. This code invoked a `Client` method, rather than `talkToApstra()`, so any error should already have been converted to a `ClientErr`.